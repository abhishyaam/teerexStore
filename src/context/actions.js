export const ADD_TO_CART = 'ADD_TO_CART';
export const REMOVE_FROM_CART = 'REMOVE_FROM_CART';
export const UPDATE_ITEM_QTY = 'UPDATE_ITEM_QTY';
export const FILTER_BY_SEARCH = 'FILTER_BY_SEARCH';
export const POPULATE_PRODUCTS = 'POPULATE_PRODUCTS';
export const POPULATE_FILTER_PARAMS = 'POPULATE_FILTER_PARAMS';
