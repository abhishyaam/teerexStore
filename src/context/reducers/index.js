import {
  ADD_TO_CART,
  REMOVE_FROM_CART,
  UPDATE_ITEM_QTY,
  POPULATE_PRODUCTS,
  FILTER_BY_SEARCH,
  POPULATE_FILTER_PARAMS,
} from '../actions';

//data reducer
export const dataReducer = (state, action) => {
  switch (action.type) {
    case ADD_TO_CART:
      return { ...state, cart: [...state.cart, { ...action.payload, qty: 1 }] };
    case REMOVE_FROM_CART:
      return {
        ...state,
        cart: state.cart.filter((item) => item.id !== action.payload.id),
      };
    case UPDATE_ITEM_QTY:
      return {
        ...state,
        cart: state.map((item) =>
          item.id === action.payload.id ? action.payload.qty : item.qty
        ),
      };

    case POPULATE_PRODUCTS:
      return {
        ...state,
        products: [...action.payload],
      };
  }
};

//filters reducer
export const filtersReducer = (state, action) => {
  switch (action.type) {
    case FILTER_BY_SEARCH:
      return { ...state, searchQuery: action.payload };
    case POPULATE_FILTER_PARAMS:
      return { ...state, filterParams: action.payload };
  }
  return state;
};
