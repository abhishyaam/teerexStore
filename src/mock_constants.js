export const mockProducts = [
  {
    color: 'Black',
    currency: 'INR',
    gender: 'Men',
    id: 1,
    imageURL:
      'https://geektrust.s3.ap-southeast-1.amazonaws.com/coding-problems/shopping-cart/polo-tshirts.png',
    name: 'Black Polo',
    price: 250,
    quantity: 3,
    type: 'Polo',
  },
  {
    color: 'Red',
    currency: 'INR',
    gender: 'Men',
    id: 2,
    imageURL:
      'https://geektrust.s3.ap-southeast-1.amazonaws.com/coding-problems/shopping-cart/polo-tshirts.png',
    name: 'Red Polo',
    price: 250,
    quantity: 3,
    type: 'Polo',
  },
  {
    color: 'Blue',
    currency: 'INR',
    gender: 'Men',
    id: 3,
    imageURL:
      'https://geektrust.s3.ap-southeast-1.amazonaws.com/coding-problems/shopping-cart/polo-tshirts.png',
    name: 'Blue Polo',
    price: 250,
    quantity: 3,
    type: 'Polo',
  },
];
