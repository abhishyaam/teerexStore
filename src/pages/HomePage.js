import React from 'react';
import { Filters, ProductItem } from '../components';
import { TeeRexState } from '../context';
import { Stack, Container, Divider, Grid } from '@mui/material';
import './styles.css';

const HomePage = () => {
  const {
    state: { products } = {},
    filtersState: {
      searchQuery,
      colorsFilter,
      typesFilter,
      gendersFilter,
      pricesFilter,
    } = {},
  } = TeeRexState();

  const filterProducts = () => {
    console.log(
      searchQuery,
      colorsFilter,
      typesFilter,
      gendersFilter,
      pricesFilter
    );
    let filteredProducts = products;
    if (colorsFilter?.length > 0) {
      filteredProducts = filteredProducts.filter(({ color }) =>
        colorsFilter.includes(color)
      );
    }

    if (typesFilter?.length > 0) {
      filteredProducts = filteredProducts.filter(({ type }) =>
        typesFilter.includes(type)
      );
    }

    if (gendersFilter?.length > 0) {
      filteredProducts = filteredProducts.filter(({ gender }) =>
        gendersFilter.includes(gender)
      );
    }

    if (pricesFilter?.length > 0) {
      filteredProducts = filteredProducts.filter(
        ({ price }) => price >= pricesFilter[0] && price <= pricesFilter[1]
      );
    }

    if (searchQuery?.length > 0) {
      filteredProducts = filteredProducts.filter((product) => {
        return (
          product.name.toLowerCase().includes(searchQuery.toLowerCase()) ||
          product.type.toLowerCase().includes(searchQuery.toLowerCase()) ||
          product.color.toLowerCase().includes(searchQuery.toLowerCase())
        );
      });
    }

    return filteredProducts;
  };

  return (
    <Container maxWidth='xl' sx={{ p: 2 }}>
      <Stack
        direction='row'
        justifyContent='flex-start'
        alignItems='flex-start'
        spacing={{ xs: 1, sm: 2, md: 4 }}
        divider={<Divider orientation='vertical' flexItem />}
      >
        <Filters />
        <Grid container spacing={2}>
          {filterProducts()?.length ? (
            filterProducts()?.map((product) => (
              <ProductItem key={product.id} product={product} />
            ))
          ) : (
            <p>Loading products...</p>
          )}
        </Grid>
      </Stack>
    </Container>
  );
};

export default HomePage;
