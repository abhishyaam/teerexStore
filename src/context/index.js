import React, { useContext, createContext, useReducer, useState } from 'react';
import { useEffect } from 'react';
import { dataReducer, filtersReducer } from './reducers';
import { POPULATE_FILTER_PARAMS, POPULATE_PRODUCTS } from './actions';
import { getProducts } from '../apis/geektrust';
import { getFilterParams } from '../utils';
/**
 * Application Context
 * Uses useReducer Hook for state managment
 *
 *
 */
const Context = createContext({
  // state: { cart: [], products: [] },
});

/**
 *
 * Store context provider
 * uses useReducer hook to provide app level data
 *
 */
const TeeRexProvider = ({ children }) => {
  const [user, setUser] = useState(null);

  const [state, dispatch] = useReducer(dataReducer, {
    products: [],
    cart: [],
  });

  const [filtersState, filtersDispatch] = useReducer(filtersReducer, {
    searchQuery: '',
    filterParams: {},
    colorsFilter: [],
    gendersFilter: [],
    typesFilter: [],
  });

  useEffect(() => {
    async function callGetProductsApi() {
      const response = await getProducts();
      const productList = await response;
      dispatch({ type: POPULATE_PRODUCTS, payload: productList });
      filtersDispatch({
        type: POPULATE_FILTER_PARAMS,
        //get filter params from a utility function
        payload: getFilterParams(productList),
      });
    }
    callGetProductsApi();
  }, []);

  return (
    <Context.Provider
      value={{ user, setUser, state, dispatch, filtersState, filtersDispatch }}
    >
      {children}
    </Context.Provider>
  );
};

export default TeeRexProvider;

export const TeeRexState = () => useContext(Context) || {};
