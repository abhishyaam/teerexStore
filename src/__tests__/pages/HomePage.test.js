import HomePage from '../../pages/HomePage';
import { getByText, render, screen } from '@testing-library/react';
import { TeeRexState } from '../../context';
import { mockProducts } from '../../mock_constants';

jest.mock('../../context', () => ({
  ...jest.requireActual('../../context'),
  TeeRexState: jest.fn(() => {}),
  filterDispatch: jest.fn(),
}));

describe('<HomePage />', () => {
  it('should render <Filters/>', () => {
    TeeRexState.mockReturnValue({
      state: { products: [] },
      filtersState: { searchQuery: '' },
    });
    const { getByText } = render(<HomePage />);
    expect(getByText('Colors')).toBeInTheDocument();
  });

  it('should display message loading products when there are no products to display', () => {
    TeeRexState.mockReturnValue({
      state: { products: null },
      filtersState: { searchQuery: '' },
    });

    render(<HomePage />);
    // console.log(render(<HomePage />))
    expect(screen.getByText('Loading products...')).toBeInTheDocument();
  });

  it('should not break when searchQuery is null', () => {
    TeeRexState.mockReturnValue({
      state: { products: mockProducts },
      filtersState: { searchQuery: null },
    });
    const { queryAllByTestId } = render(<HomePage />);
    expect(queryAllByTestId('test-product-item').length).toEqual(3);
  });

  it('search should filter products by search word', () => {
    TeeRexState.mockReturnValue({
      state: { products: mockProducts },
      filtersState: { searchQuery: 'red' },
    });
    const { queryAllByTestId } = render(<HomePage />);
    expect(queryAllByTestId('test-product-item').length).toEqual(1);
  });

  it('search ');

  it('to match snapshot', () => {
    TeeRexState.mockReturnValue({
      state: { products: mockProducts },
      filtersState: { searchQuery: null },
    });

    const { container } = render(<HomePage />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
