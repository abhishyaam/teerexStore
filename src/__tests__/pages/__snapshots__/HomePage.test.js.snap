// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> to match snapshot 1`] = `
<div
  class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
>
  <div
    class="css-4auhoz-MuiStack-root"
  >
    <div
      class="css-nbv7k5-MuiStack-root"
    >
      <div
        class="css-nen11g-MuiStack-root"
      >
        <div
          class="MuiBox-root css-qkupkn"
          elevation="0"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Colors
          </h6>
          <div
            class="css-nen11g-MuiStack-root"
          />
        </div>
        <div
          class="MuiBox-root css-qkupkn"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Types
          </h6>
          <div
            class="css-nen11g-MuiStack-root"
          />
        </div>
        <div
          class="MuiBox-root css-qkupkn"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Genders
          </h6>
          <div
            class="css-nen11g-MuiStack-root"
          />
        </div>
        <div
          class="MuiBox-root css-1mzovj4"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Price
          </h6>
          <div
            class="css-nen11g-MuiStack-root"
          >
            <span
              class="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root css-187mznn-MuiSlider-root"
            >
              <span
                class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
              />
              <span
                class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
                style="left: 0%; width: 100%;"
              />
              <span
                class="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
                data-index="0"
                style="left: 0%;"
              >
                <input
                  aria-orientation="horizontal"
                  aria-valuemax="500"
                  aria-valuemin="0"
                  aria-valuenow="0"
                  data-index="0"
                  max="500"
                  min="0"
                  step="1"
                  style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                  type="range"
                  value="0"
                />
                <span
                  aria-hidden="true"
                  class="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
                >
                  <span
                    class="MuiSlider-valueLabelCircle"
                  >
                    <span
                      class="MuiSlider-valueLabelLabel"
                    >
                      0
                    </span>
                  </span>
                </span>
              </span>
              <span
                class="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
                data-index="1"
                style="left: 100%;"
              >
                <input
                  aria-orientation="horizontal"
                  aria-valuemax="500"
                  aria-valuemin="0"
                  aria-valuenow="500"
                  data-index="1"
                  max="500"
                  min="0"
                  step="1"
                  style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
                  type="range"
                  value="500"
                />
                <span
                  aria-hidden="true"
                  class="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
                >
                  <span
                    class="MuiSlider-valueLabelCircle"
                  >
                    <span
                      class="MuiSlider-valueLabelLabel"
                    >
                      500
                    </span>
                  </span>
                </span>
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-1tk1bks-MuiDivider-root"
    />
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 MuiGrid-grid-lg-2 css-1pgzwh3-MuiGrid-root"
        data-testid="test-product-item"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1odlhgd-MuiPaper-root"
        >
          <div>
            https://geektrust.s3.ap-southeast-1.amazonaws.com/coding-problems/shopping-cart/polo-tshirts.png
          </div>
          <div
            class="css-5z1x7y-MuiStack-root"
          >
            <div>
              250
            </div>
            <div>
              Black Polo
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 MuiGrid-grid-lg-2 css-1pgzwh3-MuiGrid-root"
        data-testid="test-product-item"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1odlhgd-MuiPaper-root"
        >
          <div>
            https://geektrust.s3.ap-southeast-1.amazonaws.com/coding-problems/shopping-cart/polo-tshirts.png
          </div>
          <div
            class="css-5z1x7y-MuiStack-root"
          >
            <div>
              250
            </div>
            <div>
              Red Polo
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-4 MuiGrid-grid-md-3 MuiGrid-grid-lg-2 css-1pgzwh3-MuiGrid-root"
        data-testid="test-product-item"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1odlhgd-MuiPaper-root"
        >
          <div>
            https://geektrust.s3.ap-southeast-1.amazonaws.com/coding-problems/shopping-cart/polo-tshirts.png
          </div>
          <div
            class="css-5z1x7y-MuiStack-root"
          >
            <div>
              250
            </div>
            <div>
              Blue Polo
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
